### PR TITLE
User registration

### PR DESCRIPTION
**Pull request by ajtran303**
### New Feature: User Registration

Users can now register by filling out a form.
They must meet certain requirements:

- The form must be completely filled out
- The zip code must be a number
- The email address must be unique
- The password must be confirmed by entering it in twice.

If these requirements are met, a user can submit their form.
If not, they have to try again.
They will see a message that indicates what info is missing.
The form will keep all the other information they entered except for the email and password.

### User Model Validation

If the form data is sent as valid params to the User model, it will be recorded into the database.

The BCrypt gem is used to store the user's password as an encrypted hash digest.

Additionally, a user who is created this way will by default, have a `role` of "Regular User".

### FactoryBot Syntax for using User Factories to make users with different roles

`create(:user, role: 0)` or `build(:user)` - Regular User
`create(:user, role: 1)` - Merchant Employee User
`create(:user, role: 2)` - Admin User

The create syntax will record a new user into the testing environment database. 
The default and implicit role is Regular User (it can be explicitly regular too!).

### CODE REVIEW

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -0,0 +1,40 @@
+class UsersController < ApplicationController
+  def new
+    @user = User.new(session[:user_params])
```
Comment (ljwhite) [inline app/controllers/users_controller.rb]: Line 3: is the session coming from the create action?

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    session.delete(:user_params)
```
Comment (ljwhite) [inline app/controllers/users_controller.rb]: What is the purpose of the delete.session? Test suite passes without it.

File: app/controllers/users_controller.rb [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+  end
+
+  def create
+    user = User.new(user_params)
+    if user.valid? && passwords_match?
+      user.save
+      flash[:success] = "You are now registered and logged in!"
+      redirect_to "/profile"
+    else
+      flash[:failure] = "All fields must be completed before submitting:"
```
Comment (ljwhite) [inline app/controllers/users_controller.rb:14]: Is creating customized flash types this easy? For earlier versions of Rails, you had to insert add_flash_type into ApplicationController. That's cool - good to know.

File: app/views/users/new.html.erb [hunk 4]
```diff
@@ -0,0 +1,19 @@
+<%= form_tag "/users", method: :post, class: "new-user-form" do %>
+<%= label_tag :name, "Full Name" %>
```
Comment (ljwhite) [inline app/views/users/new.html.erb:2]: Formatting question - indent all label_tag and text_field_tag?

File: spec/features/users/new_user_registration_spec.rb [hunk 5]
```diff
@@ -0,0 +1,38 @@
+require "rails_helper"
+
+RSpec.describe "New User Registration Spec" do
+  before :each do
+    @user = build(:user)
```
Comment (ljwhite) [inline spec/features/users/new_user_registration_spec.rb:5]: FactoryBot makes this so easy. Love it.

File: spec/features/users/new_user_registration_unique_email_spec.rb [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+        click_button "Register as a New User"
+
+        expect(User.all).to_not include(@user2)
+
+        expect(current_path).to eq(register_path)
+
+        expect(page).to have_content("Email has already been taken")
+
+        expect(page).to have_selector("input[value='#{@user2.name}']")
```
Comment (ljwhite) [inline spec/features/users/new_user_registration_unique_email_spec.rb:34]: I'm not sure I understand the advantage of this over expect(page).to have_content(@user2.name)? Curious to hear more.

File: spec/features/users/new_user_registration_validation_spec.rb [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+          fill_in :city, with: @user.city
+          fill_in :state, with: @user.state
+          fill_in :zip, with: @user.zip
+          fill_in :email, with: @user.email
+          fill_in :password, with: @user.password
+          fill_in :confirm_password, with: @user.password
+
+          click_button "Register as a New User"
+
+          expect(User.all.count).to eq(0)
```
Comment (ljwhite) [inline spec/features/users/new_user_registration_validation_spec.rb:27]: Consider duplicating this test on line 24. Not necessary though.

File: app/controllers/users_controller.rb (hunk 1 above)
Comment (ajtran303) [reply app/controllers/users_controller.rb]: Yes

File: app/views/users/new.html.erb (hunk 4 above)
Comment (ajtran303) [reply app/views/users/new.html.erb:2]: I used the auto-indentations and it came out flush like this.

File: spec/features/users/new_user_registration_unique_email_spec.rb (hunk 6 above)
Comment (ajtran303) [reply spec/features/users/new_user_registration_unique_email_spec.rb:34]: I'm specifically looking in the part of the form where a user can input text -- it leaves no room for doubt.

File: app/controllers/users_controller.rb (hunk 2 above)
Comment (ajtran303) [reply app/controllers/users_controller.rb]: I only want to store that user data in the session for as long as it is needed.

If the user successfully registers, then that data is saved into the database. 
In this case, we already have it saved, so we can delete it from the sesson.

It adds an extra layer of security -> in a later commit, I actually limit the params that are passed into the sesson.

Comment (ajtran303) [reply app/controllers/users_controller.rb]:
```
  def user_registration_params
    params.permit(:name, :address, :city, :state, :zip, :email)
  end
```

**Pull request merged**